dataset/src/main/scala/frameless/TypedEncoder.scala

```diff
@@ -296,24 +337,141 @@ object TypedEncoder {
     }
   }
 
+  implicit def arrayEncoder[A : ClassTag](
+    implicit
+    underlying: TypedEncoder[A]
+  ): TypedEncoder[Array[A]] = new TypedEncoder[Array[A]]() {
+    val nullable: Boolean = false
+
+    val sourceDataType: DataType = FramelessInternals.objectTypeFor[Array[A]](classTag)
+
+    val targetDataType: DataType = DataTypes.createArrayType(underlying.targetDataType)
+
+    def constructorFor(path: Expression): Expression = {
+      Option(underlying.sourceDataType)
+        .filter(ScalaReflection.isNativeType)
+        .filter(_ == underlying.targetDataType)
+        .collect {
+          case BooleanType => "toBooleanArray" -> ScalaReflection.dataTypeFor[Array[Boolean]]
+          case ByteType => "toByteArray" -> ScalaReflection.dataTypeFor[Array[Byte]]
+          case ShortType => "toShortArray" -> ScalaReflection.dataTypeFor[Array[Short]]
+          case IntegerType => "toIntArray" -> ScalaReflection.dataTypeFor[Array[Int]]
+          case LongType => "toLongArray" -> ScalaReflection.dataTypeFor[Array[Long]]
+          case FloatType => "toFloatArray" -> ScalaReflection.dataTypeFor[Array[Float]]
+          case DoubleType => "toDoubleArray" -> ScalaReflection.dataTypeFor[Array[Double]]
+        }.map {
+        case (method, typ) => Invoke(path, method, typ)
+      }.getOrElse {
+        Invoke(
+          MapObjects(
+            underlying.constructorFor,
+            path,
+            underlying.targetDataType
+          ),
+          "array",
+          sourceDataType
+        )
+      }
+    }
+
+    def extractorFor(path: Expression): Expression = {
+      // if source `path` is already native for Spark, no need to `map`
+      if (ScalaReflection.isNativeType(underlying.sourceDataType)) {
+        NewInstance(
+          classOf[GenericArrayData],
+          path :: Nil,
+          dataType = ArrayType(underlying.targetDataType, underlying.nullable)
+        )
+      } else {
+        MapObjects(underlying.extractorFor, path, underlying.sourceDataType)
+      }
+    }
+  }
+
+  implicit def mapEncoder[A : ClassTag, B : ClassTag](
+    implicit
+    encodeA: TypedEncoder[A],
+    encodeB: TypedEncoder[B]
+  ): TypedEncoder[scala.collection.immutable.Map[A, B]] = new TypedEncoder[scala.collection.immutable.Map[A, B]] {
```

Comment:
IIRC Spark reflection encoder is for `Map` is for `scala.collection.Map` and not `scala.collection.immutable.Map`. I wonder if we should follow (or provide both) for compatibility.